db/db_impl/db_impl.cc

```diff
@@ -1264,11 +1264,14 @@ Status DBImpl::ApplyReplicationLogRecord(ReplicationLogRecord record,
             // argue: at least one alive_log_files_ and logs_ always needs to be
             // alive and alive means that its log number is greater or equal to
             // the latest log number.
-            versions_->MarkFileNumberUsed(e.GetLogNumber());
             alive_log_files_.begin()->number = e.GetLogNumber();
             logs_.front().number = e.GetLogNumber();
           }
 
+          if (e.HasNextFile()) {
+              versions_->SetNextFileNumber(e.GetNextFile());
```

Comment:
hmm I still don't quite get the issue based on the PR description, especially confused by this sentence: `Version edit's next file number (that was set on the
leader) was reset in follower's VersionSet based on the latest value of
next_file_number_ here`.

`log_number_`  should be set as latest `next_file_number` whenever we switch memtable. So what cases are we missing here? file number increase caused by compaction?

Reply:
No, file number increase caused by flush. There is no compaction in the test.

* Memtable switch, log_number=9, next_file_number=10
* Flush into file 10.sst, next_file_number=10